File: termoroyale_backend/termoroyale/src/main/java/com/fernando/termoroyale/core/usecase/TermoValidator.java
package com.fernando.termoroyale.core.usecase;

import com.fernando.termoroyale.core.domain.LetterStatus;
import org.springframework.stereotype.Component;
import java.util.HashMap;
import java.util.Map;

@Component
public class TermoValidator {

    public LetterStatus[] validate(String guess, String target) {
        int length = target.length();
        LetterStatus[] results = new LetterStatus[length];

        Map<Character, Integer> targetLetterCounts = new HashMap<>();
        for (char c : target.toUpperCase().toCharArray()) {
            targetLetterCounts.put(c, targetLetterCounts.getOrDefault(c, 0) + 1);
        }

        char[] guessChars = guess.toUpperCase().toCharArray();
        char[] targetChars = target.toUpperCase().toCharArray();

        // 1ª Passagem: VERDE (CORRECT)
        for (int i = 0; i < length; i++) {
            if (guessChars[i] == targetChars[i]) {
                results[i] = LetterStatus.CORRECT;
                targetLetterCounts.put(guessChars[i], targetLetterCounts.get(guessChars[i]) - 1);
            }
        }

        // 2ª Passagem: AMARELO (PRESENT) ou CINZA (ABSENT)
        for (int i = 0; i < length; i++) {
            if (results[i] == null) {
                char c = guessChars[i];
                if (targetLetterCounts.getOrDefault(c, 0) > 0) {
                    results[i] = LetterStatus.PRESENT;
                    targetLetterCounts.put(c, targetLetterCounts.get(c) - 1);
                } else {
                    results[i] = LetterStatus.ABSENT;
                }
            }
        }
        return results;
    }
}

File: termoroyale_frontend/termo-royale-web-ts/src/utils/termoValidator.ts
import type { LetterStatus } from "../types/game";

/**
 * Replica em TypeScript do TermoValidator (Java).
 * Marca CORRECT em 1ª passagem, depois PRESENT/ABSENT respeitando contagem de letras duplicadas.
 */
export function validateGuess(guess: string, target: string): LetterStatus[] {
    const g = guess.toUpperCase();
    const t = target.toUpperCase();
    const length = t.length;
    const result: LetterStatus[] = new Array(length).fill("ABSENT");

    const counts: Record<string, number> = {};
    for (const c of t) counts[c] = (counts[c] ?? 0) + 1;

    // 1ª passagem: verde
    for (let i = 0; i < length; i++) {
        if (g[i] === t[i]) {
            result[i] = "CORRECT";
            counts[g[i]]--;
        }
    }

    // 2ª passagem: amarelo / cinza
    for (let i = 0; i < length; i++) {
        if (result[i] === "CORRECT") continue;
        const c = g[i];
        if ((counts[c] ?? 0) > 0) {
            result[i] = "PRESENT";
            counts[c]--;
        } else {
            result[i] = "ABSENT";
        }
    }
    return result;
}

/** Calcula key statuses agregados a partir de todas as tentativas (igual ao multiplayer). */
export function computeKeyStatuses(
    guesses: string[],
    results: LetterStatus[][]
): Record<string, LetterStatus> {
    const out: Record<string, LetterStatus> = {};
    guesses.forEach((word, gi) => {
        const row = results[gi];
        if (!row) return;
        word.split("").forEach((letter, li) => {
            const cur = row[li];
            if (out[letter] === "CORRECT") return;
            if (out[letter] === "PRESENT" && cur === "ABSENT") return;
            out[letter] = cur;
        });
    });
    return out;
}
